## Association Rule Mining in Retail Store

### Problem Statement:
 * What are the items that may be frequently purchased together?

### Objective:
* To learn how Apriori Algorithm and Association Rules works.
* To learn how Combination and Permutation helps to find Support and Confidence of itemsets respectively.
* To find frequent itemsets with high confidence and lift, keeping both item together will help to increase sales.


### Introduction
* Association rule mining is one of an important technique of data mining for knowledge discovery.
* The knowledge of the correlation between the items in the data transaction can use association rule mining.
* Retail store analysis is one of an application area of association rule mining technique.
* The possible percentage of the correlation of combined items gives the new knowledge. Therefore, it is a very helpful for determiner to take the decisions

### Analysis

In [1]:
# Importing Required Library

import pandas as pd
import numpy as np

In [2]:
# Reading Excel file 
# need to install 1.2.0 version xlrd package to read excel files: pip install xlrd==1.2.0
bread = pd.read_excel('raw_bread.xlsx')

In [3]:
## Here we have transaction data, which include column, Date,Time,Transaction,Item
## we should remove duplicate transaction, it shows quantity of item in same transaction,
## it is not needed in appriori aglo as we only care about different item in particular transaction
bread

Date,Time,Transaction,Item
0              2016-10-30,09:58:11,1,Bread
1       2016-10-30,10:05:34,2,Scandinavian
2       2016-10-30,10:05:34,2,Scandinavian
3      2016-10-30,10:07:57,3,Hot chocolate
4                2016-10-30,10:07:57,3,Jam
...                                    ...
21288      2017-04-09,14:32:58,9682,Coffee
21289         2017-04-09,14:32:58,9682,Tea
21290      2017-04-09,14:57:06,9683,Coffee
21291      2017-04-09,14:57:06,9683,Pastry
21292   2017-04-09,15:04:24,9684,Smoothies

[21293 rows x 1 columns]

In [4]:
## dropping Duplicate Transaction
bread = bread.drop_duplicates()

In [5]:
## we need to split transaction data into Dataframe/tabular structure as follow
new = bread['Date,Time,Transaction,Item'].str.split(',', n = 3, expand = True)

In [6]:
import warnings
warnings.filterwarnings('ignore')

In [7]:
## assigning column to data frame "bread"
bread['Date'] = new[0]
bread['Time'] = new[1]
bread['Transaction'] = new[2]
bread['Item'] = new[3]

In [8]:
# in this dataframe we only need column Trasaction and Item, rest is not needed in association mining rule
bread[['Date', 'Time', 'Transaction', 'Item']].head(10)
                                                    

Date      Time Transaction           Item
0   2016-10-30  09:58:11           1          Bread
1   2016-10-30  10:05:34           2   Scandinavian
3   2016-10-30  10:07:57           3  Hot chocolate
4   2016-10-30  10:07:57           3            Jam
5   2016-10-30  10:07:57           3        Cookies
6   2016-10-30  10:08:41           4         Muffin
7   2016-10-30  10:13:03           5         Coffee
8   2016-10-30  10:13:03           5         Pastry
9   2016-10-30  10:13:03           5          Bread
10  2016-10-30  10:16:55           6      Medialuna

In [9]:
# we need to convert cloumn transacton & item into Crosstab or we can say Binary Matrix as follow
transaction = pd.crosstab(index= bread['Transaction'], columns= bread['Item'])
transaction

Item         Adjustment  Afternoon with the baker  Alfajores  Argentina Night  \
Transaction                                                                     
1                     0                         0          0                0   
10                    0                         0          0                0   
100                   0                         0          0                0   
1000                  0                         0          0                0   
1001                  0                         0          0                0   
...                 ...                       ...        ...              ...   
995                   0                         0          0                0   
996                   0                         0          0                0   
997                   0                         0          0                0   
998                   0                         0          0                0   
999                   0                         0          1                0   

Item         Art Tray  Bacon  Baguette  Bakewell  Bare Popcorn  Basket  ...  \
Transaction                                                             ...   
1                   0      0         0         0             0       0  ...   
10                  0      0         0         0             0       0  ...   
100                 0      0         0         0             0       0  ...   
1000                0      0         0         0             0       0  ...   
1001                0      0         0         0             0       0  ...   
...               ...    ...       ...       ...           ...     ...  ...   
995                 0      0         0         0             0       0  ...   
996                 0      0         0         0             0       0  ...   
997                 0      0         0         0             0       0  ...   
998                 0      0         0         0             0       0  ...   
999                 0      0         0         0             0       0  ...   

Item         The BART  The Nomad  Tiffin  Toast  Truffles  Tshirt  \
Transaction                                                         
1                   0          0       0      0         0       0   
10                  0          0       0      0         0       0   
100                 0          0       0      0         0       0   
1000                0          0       0      0         0       0   
1001                0          0       0      0         1       0   
...               ...        ...     ...    ...       ...     ...   
995                 0          0       0      0         0       0   
996                 0          0       0      0         1       0   
997                 0          0       0      0         0       0   
998                 0          0       0      0         0       0   
999                 0          0       0      0         0       0   

Item         Valentine's card  Vegan Feast  Vegan mincepie  Victorian Sponge  
Transaction                                                                   
1                           0            0               0                 0  
10                          0            0               0                 0  
100                         0            0               0                 0  
1000                        0            0               0                 0  
1001                        0            0               0                 0  
...                       ...          ...             ...               ...  
995                         0            0               0                 0  
996                         0            0               0                 0  
997                         0            0               0                 0  
998                         0            0               0                 0  
999                         0            0               0              

In [10]:
## Just writing csv file to check result
## we have one unwanted column named "NONE", we should remove it as follow and proceed further
# tab.to_csv('tab.csv')

In [11]:
## removing unwanted col "NONE"
transaction = transaction.drop(['NONE'], axis = 1)

In [12]:
transaction

Item         Adjustment  Afternoon with the baker  Alfajores  Argentina Night  \
Transaction                                                                     
1                     0                         0          0                0   
10                    0                         0          0                0   
100                   0                         0          0                0   
1000                  0                         0          0                0   
1001                  0                         0          0                0   
...                 ...                       ...        ...              ...   
995                   0                         0          0                0   
996                   0                         0          0                0   
997                   0                         0          0                0   
998                   0                         0          0                0   
999                   0                         0          1                0   

Item         Art Tray  Bacon  Baguette  Bakewell  Bare Popcorn  Basket  ...  \
Transaction                                                             ...   
1                   0      0         0         0             0       0  ...   
10                  0      0         0         0             0       0  ...   
100                 0      0         0         0             0       0  ...   
1000                0      0         0         0             0       0  ...   
1001                0      0         0         0             0       0  ...   
...               ...    ...       ...       ...           ...     ...  ...   
995                 0      0         0         0             0       0  ...   
996                 0      0         0         0             0       0  ...   
997                 0      0         0         0             0       0  ...   
998                 0      0         0         0             0       0  ...   
999                 0      0         0         0             0       0  ...   

Item         The BART  The Nomad  Tiffin  Toast  Truffles  Tshirt  \
Transaction                                                         
1                   0          0       0      0         0       0   
10                  0          0       0      0         0       0   
100                 0          0       0      0         0       0   
1000                0          0       0      0         0       0   
1001                0          0       0      0         1       0   
...               ...        ...     ...    ...       ...     ...   
995                 0          0       0      0         0       0   
996                 0          0       0      0         1       0   
997                 0          0       0      0         0       0   
998                 0          0       0      0         0       0   
999                 0          0       0      0         0       0   

Item         Valentine's card  Vegan Feast  Vegan mincepie  Victorian Sponge  
Transaction                                                                   
1                           0            0               0                 0  
10                          0            0               0                 0  
100                         0            0               0                 0  
1000                        0            0               0                 0  
1001                        0            0               0                 0  
...                       ...          ...             ...               ...  
995                         0            0               0                 0  
996                         0            0               0                 0  
997                         0            0               0                 0  
998                         0            0               0                 0  
999                         0            0               0              

### Creating APRIORI function to generate frequent itesets based on minimum threshold support = 0.04

In [83]:
def APRIORI(data, min_support = 0.04,  max_length = 4):
    # Collecting Required Library
    import numpy as np
    import pandas as pd
    import itertools
    
    support = {} 
    L = list(data.columns)  # 所有商品
    total_count = len(data)
    
    # write your code here !!
    # 先找出frequent item
    for item in L:
        item_support = len(data[data[item]==1]) / total_count
        if item_support >= 0.04:
            support[item] = item_support

    for i in range(2, max_length+1):
        # 收集frequent item
        itemsets = []
        for frequent_item in support:
            if i > 2 and type(frequent_item) == tuple:
                for item in frequent_item:
                    if item not in itemsets:
                        itemsets.append(item)
            else:
                itemsets.append(frequent_item)
                
        # 找出所有可能的組合
        itemsets = itertools.combinations(itemsets, i)
        
        for itemset in itemsets:
            # 找出符合itemset的交易
            tempdata = data
            for item in itemset:
                tempdata = tempdata[tempdata[item]==1]
                
            # 計算此itemset之support
            itemset_support = len(tempdata)/ total_count
            if itemset_support >= 0.04:
                support[itemset] = itemset_support

    result = pd.DataFrame(list(support.items()), columns = ["Items", "Support"])
    return result

In [84]:
## finding frequent itemset with min support = 4%
my_freq_itemset = APRIORI(transaction, 0.04, 3)
my_freq_itemset.sort_values(by = 'Support', ascending = False)

Items   Support
2             Coffee  0.475081
0              Bread  0.324940
8                Tea  0.141643
1               Cake  0.103137
9    (Bread, Coffee)  0.089393
6             Pastry  0.085510
7           Sandwich  0.071346
5          Medialuna  0.061379
4      Hot chocolate  0.057916
10    (Cake, Coffee)  0.054349
3            Cookies  0.054034
12     (Coffee, Tea)  0.049523
11  (Coffee, Pastry)  0.047214

In [104]:
x = pd.Series(my_freq_itemset.Support.values, index=my_freq_itemset.Items).to_dict()

[1, 2]

### Creating ASSOCIATION_RULE function to generate rules based on minimun threshold confidence.

In [115]:
def ASSOCIATION_RULE(df, min_confidence_threshold = 0.5):
    import pandas as pd
    import itertools
    
    support = pd.Series(df.Support.values, index=df.Items).to_dict()
    data = []
    L = df.Items.values
    total_count = len(data)

    # write your code here !!
    for itemset in L:
        if type(itemset) == tuple:
            total_support = support[itemset]
            itemset_list = itertools.permutations(itemset)
            for subset in itemset_list:
                ant = subset[0]
                confidence = total_support/support[ant]
                if confidence > min_confidence_threshold:
                    con = subset[1]
                    ant_support = support[ant]
                    con_support = support[con]
                    lift = confidence/con_support
                    leverage = total_support - ant_support * con_support
                    convection = (1-con_support)/(1-confidence)
                    data.append([ant, con, ant_support, con_support, total_support, confidence, lift, leverage, convection])

    result = pd.DataFrame(data, columns = ["antecedents", "consequents", "antecedent support", "consequent support",
                                        "support", "confidence", "Lift", "Leverage", "Convection"])
    return result

In [116]:
## Rule with minimun confidence = 50%
my_rule = ASSOCIATION_RULE(my_freq_itemset, 0.5)
my_rule
# the column names in the following table is the same as those generated by mlxtend
# you can refer to http://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/ 
# to find the formulas of lift, leverage, convection

antecedents consequents  antecedent support  consequent support   support  \
0        Cake      Coffee            0.103137            0.475081  0.054349   
1      Pastry      Coffee            0.085510            0.475081  0.047214   

   confidence      Lift  Leverage  Convection  
0    0.526958  1.109196   0.00535    1.109667  
1    0.552147  1.162216   0.00659    1.172079

### Finally sorting results by Lift to get highly associated itemsets.

In [117]:
my_rule.sort_values(by='Lift', ascending= False).head(10)

antecedents consequents  antecedent support  consequent support   support  \
1      Pastry      Coffee            0.085510            0.475081  0.047214   
0        Cake      Coffee            0.103137            0.475081  0.054349   

   confidence      Lift  Leverage  Convection  
1    0.552147  1.162216   0.00659    1.172079  
0    0.526958  1.109196   0.00535    1.109667

## Cross Verifying results with  apriori and association rule from mlxtend

In [20]:
# Loading standard package
from mlxtend.frequent_patterns import apriori, association_rules

In [21]:
## finding frequent itemset with min support = 4%
frequent_itemset = apriori(df = transaction, min_support= 0.04, use_colnames= True)
frequent_itemset.sort_values(by = 'support', ascending = False)

support          itemsets
2   0.475081          (Coffee)
0   0.324940           (Bread)
8   0.141643             (Tea)
1   0.103137            (Cake)
9   0.089393   (Coffee, Bread)
6   0.085510          (Pastry)
7   0.071346        (Sandwich)
5   0.061379       (Medialuna)
4   0.057916   (Hot chocolate)
10  0.054349    (Coffee, Cake)
3   0.054034         (Cookies)
12  0.049523     (Coffee, Tea)
11  0.047214  (Coffee, Pastry)

### Createing associate rule such that item brought with conditional probability(Confidence) more than 50% with corresponding item

In [22]:
## Rule with minimun confidence = 50%
Rules = association_rules(frequent_itemset, min_threshold= 0.5)
Rules
[['Cake', 'Coffee', 0.10313713146574337, 0.4750813136082258, 0.5269582909460834, 0.11439929328621908, 0.0053504426317400045, 0.5550870964162876], 
 ['Pastry', 'Coffee', 0.08551043961808834, 0.4750813136082258, 0.5521472392638037, 0.09938162544169611, 0.006589941182383381, 0.550930514260544]]

antecedents consequents  antecedent support  consequent support   support  \
0      (Cake)    (Coffee)            0.103137            0.475081  0.054349   
1    (Pastry)    (Coffee)            0.085510            0.475081  0.047214   

   confidence      lift  leverage  conviction  
0    0.526958  1.109196   0.00535    1.109667  
1    0.552147  1.162216   0.00659    1.172079

In [23]:
## Finally sorting results by Lift to get highly associated itemsets.
Rules.sort_values(by='lift', ascending= False).head(10)

antecedents consequents  antecedent support  consequent support   support  \
1    (Pastry)    (Coffee)            0.085510            0.475081  0.047214   
0      (Cake)    (Coffee)            0.103137            0.475081  0.054349   

   confidence      lift  leverage  conviction  
1    0.552147  1.162216   0.00659    1.172079  
0    0.526958  1.109196   0.00535    1.109667

### Conclusion
 * Results from developed function(APRIORI, ASSOCIATION_RULE) has matched with builts packages.
 * it is observed that "Toast" & "Coffee" are highly associated with lift 1.48.
 * Coffee has been brought most frequently with 47.5% of all the transactions